bridge/src/send_tx_with_receipt.rs

```diff
@@ -0,0 +1,368 @@
+                                .poll()
+                                .chain_err(|| "SendTransactionWithReceipt: getting transaction receipt failed")
+                        );
+
+                        match maybe_receipt {
+                            // transaction hasn't been mined yet
+                            None => State::AwaitBlockNumber(transaction_hash),
+                            Some(receipt) => {
+                                info!("SendTransactionWithReceipt: got transaction receipt: {}", transaction_hash);
+                                match receipt.block_number {
+                                    // receipt comes from pending block
+                                    None => State::AwaitBlockNumber(transaction_hash),
+                                    Some(receipt_block_number) => {
+                                        if last_block < receipt_block_number.as_u64() {
+                                            // transaction does not have enough confirmations
+                                            State::AwaitBlockNumber(transaction_hash)
+                                        } else {
+                                            return Ok(Async::Ready(receipt))
+                                        }
+                                    }
+                                }
+                            },
+                        }
+                    }
+                };
+
+                self.state = next_state;
+            }
+        }
+    }
+}
+
+enum State<T: Transport> {
+    AwaitBlockNumber {
+        future: Timeout<FromErr<CallFuture<U256, T::Out>, error::Error>>,
+        transaction: Option<TransactionRequest>,
+    },
+    AwaitReceipt(inner::SendTransactionWithReceipt<T>)
+}
+
+
+pub struct SendTransactionWithReceiptOptions<T> {
+    pub transport: T,
+    pub request_timeout: Duration,
+    pub poll_interval: Duration,
+    pub confirmations: u32,
+    pub transaction: TransactionRequest,
+}
+
+pub struct SendTransactionWithReceipt<T: Transport> {
+    request_timeout: Duration,
+    poll_interval: Duration,
+    transport: T,
+    state: State<T>,
+    confirmations: u32,
+}
+
+impl<T: Transport> SendTransactionWithReceipt<T> {
+    pub fn new(options: SendTransactionWithReceiptOptions<T>) -> Self {
+        let timer = Timer::default();
+
+        let future = web3::api::Eth::new(&options.transport).block_number();
+
+        let state = State::AwaitBlockNumber {
+            future: timer.timeout(future.from_err(), options.request_timeout),
+            transaction: Some(options.transaction),
+        };
+
+        SendTransactionWithReceipt {
+            request_timeout: options.request_timeout,
+            poll_interval: options.poll_interval,
+            transport: options.transport,
+            state,
+            confirmations: options.confirmations,
+        }
+    }
+}
+
+impl<T: Transport> Future for SendTransactionWithReceipt<T> {
```

Comment:
Couldn't this be done simply with `.then` combinator of future?
i.e:
```
block_number_stream.into_future().then(|(block_number, _)| {
  inner::SendTransactionWithReceipt(inner::SendTransactionWithReceiptOptions {
   ...
  })
})

Reply:
hmm.. maybe it could be simplified with `web3::api::Eth::new(&transport).block_number().and_then(|(block_number)`  if also use `impl trait` as a result type. But at the same time, I kind of find this more explicit implementation to be cleaner and more descriptive :)